driver/random.hpp

```diff
@@ -1,32 +1,93 @@
 #ifndef GUARD_RANDOM_GEN_
 #define GUARD_RANDOM_GEN_
 
+#include <miopen/env.hpp>
+#include <iostream>
 #include <random>
 
-std::minstd_rand& get_minstd_gen()
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_DRIVER_PRNG_SEED)
+namespace prng {
+namespace details {
+using glibc_gen = std::linear_congruential_engine<std::uint32_t, 1103515245, 12345, 2147483648>;
+
+inline std::random_device::result_type get_default_seed()
+{
+    static std::random_device::result_type seed{[] {
+        auto external_seed = miopen::Value(MIOPEN_DEBUG_DRIVER_PRNG_SEED{}, 100500);
+
+        auto seed = external_seed == 0
+                        ? std::random_device{}()
+                        : static_cast<std::random_device::result_type>(external_seed);
+        std::cout << "PRNG seed: " << seed << "\n";
+        return seed;
+    }()};
+    return seed;
+}
+
+inline glibc_gen& get_prng()
+{
+    static thread_local glibc_gen gen{get_default_seed()};
+    return gen;
+}
+} // namespace details
+
+inline void reset_seed(std::random_device::result_type seed = 0)
 {
-    static thread_local std::minstd_rand minstd_gen(std::random_device{}());
-    return minstd_gen;
+    details::get_prng().seed(seed + details::get_default_seed());
 }
 
+// similar to std::generate_canonical, but simpler and faster
 template <typename T>
-inline T FRAND()
+inline T gen_canonical()
 {
-    double d = std::generate_canonical<double, 1>(get_minstd_gen());
-    return static_cast<T>(d);
+    if constexpr(std::is_floating_point_v<T>) // native fp
+    {
+        static constexpr T range =
+            static_cast<T>(1) /
+            static_cast<T>(details::glibc_gen::max() - details::glibc_gen::min() + 1);
+        return range * static_cast<T>(details::get_prng()() - details::glibc_gen::min());
+    }
+    else if constexpr(std::is_integral_v<T>)
+    {
+        auto val = details::get_prng()();
+        return static_cast<T>(((val >> 4) ^ (val >> 16)) & 0x1);
+    }
+    else
+    {
+        return static_cast<T>(gen_canonical<float>());
+    }
 }
 
-inline int GET_RAND()
+template <typename T>
+inline T gen_0_to_B(T B)
```

Comment:
[Q][quality] This function is very sensitive to the type of B. For example, `static_cast<half>gen_0_to_B(4.)` and `static_cast<hald>gen_0_to_B(4)` would yield totally different results. I am not sure this is good.

Maybe it’s worth making a couple of functions `gen_...` and `gen_integer_...` and removing conditional statements from the function body at once?

Reply:
Probably. It will be more explicit.
But we can't get rid of conditional statement since we can't generate half and bfloat directly.
Or we can?

Reply:
And it will require runtime checks in the callee code.
For example:
```
template <Tgpu> run()
{
    Tgpu range = is_half ? Tgpu(4) : Tgpu(16);
    auto val = gen_0_to_B(range);
}
```

Reply:
@CAHEK7 we can provide explicit specializations for each (or for some) of the supported types, thus reducing the need in conditionals, right?

Reply:
`[u]int.._t`, `size_t`, `float`, `double`, `half`, `bfloat`. For me it seems that 2 conditions is better to understand and reduce code duplication.
`if constexpr` is a modern way to do more flexible specialization and guaranteed to be performed at compile time.

Reply:
@CAHEK7 I am not against using modern stuff ;) But what about the original problem? Does it seem reasonable to split into `gen_...` and `gen_integer_...` or not? I will agree with everything you suggest ;)

Reply:
It doesn't, because it breaks high-level templates.
Instead of passing templated parameters down to the call tree, each place must check the type and explicitly call appropriate function.
Like:
```
template <Tgpu> run()
{
    Tgpu range = is_half ? Tgpu(4) : Tgpu(16);
    auto val = gen_0_to_B(range);
}
```
and
```
template <Tgpu> run()
{
    Tgpu range = is_half ? Tgpu(4) : Tgpu(16);
    auto val;
    if constexpr (std::is_integer_v<Tgpu>)
        val = gen_integer_0_to_B(range);
    else
        val = gen_real_0_to_B(range);
}
```
And that must be done everywhere in the codebase instead of inside the lowest level function.

Reply:
[Resolved]